test/unit/cmap/auth/mongodb_oidc/token_entry_cache.test.ts

```diff
@@ -0,0 +1,143 @@
+
+      it('adds the server result', function () {
+        expect(entry.serverResult).to.deep.equal(serverResult);
+      });
+
+      it('creates an expiration', function () {
+        expect(entry.expiration).to.be.above(Date.now());
+      });
+    });
+
+    context('when expiresInSeconds is not provided', function () {
+      const cache = new TokenEntryCache();
+      let entry;
+
+      const expiredResult = { accessToken: 'test' };
+
+      before(function () {
+        cache.addEntry('localhost', 'user', expiredResult, serverResult);
+        entry = cache.getEntry('localhost', 'user');
+      });
+
+      it('sets an immediate expiration', function () {
+        expect(entry.expiration).to.be.below(Date.now() + 1);
+      });
+    });
+
+    context('when expiresInSeconds is null', function () {
+      const cache = new TokenEntryCache();
+      let entry;
+
+      const expiredResult = {
+        accessToken: 'test',
+        expiresInSeconds: null
+      };
+
+      before(function () {
+        cache.addEntry('localhost', 'user', expiredResult, serverResult);
+        entry = cache.getEntry('localhost', 'user');
+      });
+
+      it('sets an immediate expiration', function () {
+        expect(entry.expiration).to.be.below(Date.now() + 1);
+      });
+    });
+  });
+
+  describe('#clear', function () {
+    const cache = new TokenEntryCache();
+
+    before(function () {
+      cache.addEntry('localhost', 'user', tokenResult, serverResult);
+      cache.clear();
+    });
+
+    it('clears the cache', function () {
+      expect(cache.entries.size).to.equal(0);
+    });
+  });
+
+  describe('#deleteExpiredEntries', function () {
+    const cache = new TokenEntryCache();
+
+    const nonExpiredResult = {
+      accessToken: 'test',
+      expiresInSeconds: 600
+    };
+
+    before(function () {
+      cache.addEntry('localhost', 'user', tokenResult, serverResult);
+      cache.addEntry('localhost', 'user2', nonExpiredResult, serverResult);
+      cache.deleteExpiredEntries();
+    });
+
+    it('deletes all expired tokens from the cache', function () {
+      expect(cache.entries.size).to.equal(1);
+    });
+  });
+
+  describe('#deleteEntry', function () {
+    const cache = new TokenEntryCache();
+
+    before(function () {
+      cache.addEntry('localhost', 'user', tokenResult, serverResult);
+      cache.deleteEntry('localhost', 'user');
+    });
+
+    it('deletes the entry', function () {
+      expect(cache.entries.has('localhost-user')).to.be.false;
+    });
+  });
+
+  describe('#getEntry', function () {
+    const cache = new TokenEntryCache();
+
+    before(function () {
+      cache.addEntry('localhost', 'user', tokenResult, serverResult);
+      cache.addEntry('localhost', 'user2', tokenResult, serverResult);
+    });
+
+    context('when there is a matching entry', function () {
+      it('returns the entry', function () {
+        expect(cache.getEntry('localhost', 'user').tokenResult).to.deep.equal(tokenResult);
```

Comment:
do we also expect it to plain `equal` the reference? looking at the code, it looks like the cache stores and returns the references directly as opposed to clones (I assume that's intentional); I think it may be worthwhile to make the `tokenResult` readonly for the test so that we can indirectly assert that the cache methods do not mutate the entry; we could also add an explicit test for it to each method

Reply:
I've changed the test to check for equal by reference and also frozen all the objects in the suite to ensure they are not mutated.